### PR TITLE
Tickets/dm 33376

### PR DESCRIPTION
**Pull request by iagaponenko**

### CODE REVIEW

File: src/replica/Controller.cc [hunk 1]
```diff
@@ -83,7 +88,68 @@ ostream& operator <<(ostream& os, ControllerIdentity const& identity) {
...
+    weak_ptr<Controller> w = ptr;
+    auto const tracker = [controller=move(w), context] () {
+        LOG_LOGGER log_ = LOG_GET("lsst.qserv.replica.Controller");
+        LOGS(log_, LOG_LVL_INFO, context << "started tracking workers.");
+        while (true) {
+            Controller::Ptr const ptr = controller.lock();
+            if (ptr == nullptr) break;
+
+            auto const config = ptr->serviceProvider()->config();
+            bool const autoRegisterWorkers = config->get<unsigned int>("controller", "auto-register-workers") != 0;
```
Comment (andy-slac) [inline src/replica/Controller.cc:115]: Will there be `auto-unregister-worker`?

File: src/replica/Controller.cc [hunk 2]
```diff
@@ -142,7 +142,7 @@ Controller::Ptr Controller::create(ServiceProvider::Ptr const& serviceProvider)
                 }
             }
             this_thread::sleep_for(
-                    chrono::seconds(min(1U, config->get<unsigned int>("redirector", "heartbeat-ival-sec"))));
+                    chrono::seconds(max(1U, config->get<unsigned int>("redirector", "heartbeat-ival-sec"))));
```
Comment (andy-slac) [inline src/replica/Controller.cc:145]: Squash this commit with previous one.

File: src/replica/Messenger.cc [hunk 3]
```diff
@@ -70,16 +71,35 @@ void Messenger::cancel(string const& worker, string const& id) {
...
-    return _workerConnector.at(worker);
+MessengerConnector::Ptr const& Messenger::_connector(string const& worker) {
+    util::Lock lock(_mtx, _context(worker));
+
+    if (auto const itr = _workerConnector.find(worker); itr != _workerConnector.end()) {
+        return itr->second;
+    }
+
+    // The worker could be just added to the Configuration. In this case
+    // worker connector needs to be created nd registered in the local collection.
```
Comment (andy-slac) [inline src/replica/Messenger.cc:87]: nd -> and

File: src/replica/ConfigWorker.h [hunk 4]
```diff
@@ -148,6 +148,9 @@ class WorkerInfo {
 
     /// @return JSON representation of the object
     nlohmann::json toJson() const;
+
+    /// @return 'true' if workers objects don't have the same values of attributes
+    bool operator !=(WorkerInfo const& other) const;
```
Comment (andy-slac) [inline src/replica/ConfigWorker.h:135]: Probably makes sense to add `operator==` to have a consistent interface?

File: src/replica/ControllerApp.cc [hunk 5]
```diff
@@ -747,267 +745,175 @@ void ControllerApp::_configureParserCommandSERVICE() {
...
     if ("REPLICATE" == _requestType) {
-
         request = controller->replicate(
-            _workerName, _sourceWorkerName, _databaseName, _chunkNumber,
-            [] (Request::Ptr const& request_) { request_->print(); },
-            _priority, not _doNotTrackRequest, _allowDuplicates
-        );
+                _workerName, _sourceWorkerName, _databaseName, _chunkNumber,
+                [] (Request::Ptr const& request_) { request_->print(); },
+                _priority, !_doNotTrackRequest, _allowDuplicates);
```
Comment (andy-slac) [inline src/replica/ControllerApp.cc:755]: Did you do this with `clang-format`?

File: src/replica/IndexApp.cc [hunk 6]
```diff
@@ -114,13 +116,6 @@ IndexApp::IndexApp(int argc, char* argv[])
         "qserv-czar-db",
         "A connection URL to the MySQL server of the Qserv master database.",
         _qservCzarDbUrl
-    ).option(
-        "worker-response-timeout",
-        "Maximum timeout (seconds) to wait before the index data extraction requests sent"
-        " to workers will finish. Setting this timeout to some reasonably low number would"
-        " prevent the application from hanging for a substantial duration of time (which"
-        " depends on the default Configuration) in case if some workers were down.",
-        _timeoutSec
```
Comment (andy-slac) [inline src/replica/IndexApp.cc:123]: Did you want to remove `_timeoutSec` from header?

File: src/replica/MySQLTestApp.cc [hunk 7]
```diff
@@ -315,26 +293,21 @@ void MySQLTestApp::_executeQueryWait(string const& query) const {
...
     } else {
-
         // Note a little optimization in which the algorithm determines the file
-        // size and pre-allocates the string buffer before
+        // size and pre-allocates the string  buffer before
         // performing the actual read.
-
         ifstream fs(_fileName);
-        if (not fs) {
+        if (!fs) {
```
Comment (andy-slac) [inline src/replica/MySQLTestApp.cc:304]: Is `BANG fs` really better than `not fs`?

**Review by andy-slac — APPROVED** [review]
Approved, but with one correctable violation - please move long lambda out of `Controller::create` method, separate method is a better place for that code. Small bunch of other comments.

File: src/admin/python/lsst/qserv/admin/cli/script.py [hunk 8]
```diff
@@ -735,22 +728,22 @@ def set_initial_configuration(workers: Sequence[str]) -> None:
         block=True,
     )
 
-    # The replication controller depends on this folder existing and does not create it if it's missing.
-    # It should get fixed in DM-30074. For now we create it here.
-    os.makedirs("/qserv/data/ingest", exist_ok=True)
+    # N.B. When the controller smigs the replication database, if it is migrating from Uninitialized
```
Comment (n8pease) [inline src/admin/python/lsst/qserv/admin/cli/script.py:723]: controller -> replication controller
right?

**Pull request merged**